File: src/main/java/io/github/paul1365972/metay/Metay.java
package io.github.paul1365972.metay;

import io.github.paul1365972.metay.listener.BlockListener;
import io.github.paul1365972.metay.listener.LoadingListener;
import io.github.paul1365972.metay.storage.MetayDataStore;
import io.github.paul1365972.metay.storage.TransformingDataStore;
import io.github.paul1365972.metay.storage.endpoints.FileChunkedDataStore;
import io.github.paul1365972.metay.storage.endpoints.FolderDataStore;
import io.github.paul1365972.metay.storage.endpoints.PDCDataStore;
import io.github.paul1365972.metay.util.MCDataStoreUtil;
import org.bukkit.Chunk;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.event.Listener;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.ServicePriority;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.File;
import java.util.Arrays;

public final class Metay extends JavaPlugin implements MetayService, Listener {
	
	private MetayDataStore<Location> blockStore = null;
	private MetayDataStore<Chunk> chunkStore = null;
	private MetayDataStore<World> worldStore = null;
	private MetayDataStore<Entity> entityStore = null;
	private MetayDataStore<ItemStack> itemStore = null;
	
	@Override
	public void onLoad() {
		//TODO Add CacheDataStore to pipeline
		//TODO Estimate good cache sizes
		blockStore = new FileChunkedDataStore<>(new File(getDataFolder(), "block"), 1024,
				MCDataStoreUtil::toChunkKey, MCDataStoreUtil::toBlockKey);
		chunkStore = new FileChunkedDataStore<>(new File(getDataFolder(), "chunk"), 1024,
				MCDataStoreUtil::toSuperChunkKey, MCDataStoreUtil::toChunkKey);
		worldStore = new FolderDataStore<>(new File(getDataFolder(), "world"), world -> world.getUID().toString());
		entityStore = new TransformingDataStore<>(new PDCDataStore(), entity -> entity);
		itemStore = new TransformingDataStore<>(new PDCDataStore(), ItemStack::getItemMeta);
	}
	
	@Override
	public void onEnable() {
		getServer().getPluginManager().registerEvents(new LoadingListener(this), this);
		getServer().getPluginManager().registerEvents(new BlockListener(this), this);
		getServer().getServicesManager().register(MetayService.class, this, this, ServicePriority.Normal);
	}
	
	@Override
	public void onDisable() {
		Arrays.asList(blockStore, chunkStore, worldStore, entityStore, itemStore)
				.forEach(MetayDataStore::onClose);
	}
	
	public MetayDataStore<Location> getBlockStore() {
		return blockStore;
	}
	
	public MetayDataStore<Chunk> getChunkStore() {
		return chunkStore;
	}
	
	public MetayDataStore<World> getWorldStore() {
		return worldStore;
	}
	
	public MetayDataStore<Entity> getEntityStore() {
		return entityStore;
	}
	
	public MetayDataStore<ItemStack> getItemStore() {
		return itemStore;
	}
	
	public void onWorldLoad(World world) {
	}
	
	public void onWorldUnload(World world) {
	}
	
	public void onChunkLoad(Chunk chunk) {
	}
	
	public void onChunkUnload(Chunk chunk) {
	}
	
	public void onEntityLoad(Entity entity) {
	}
	
	public void onEntityUnload(Entity entity) {
	}
	
	public void onPlayerLoad(Player player) {
	}
	
	public void onPlayerUnload(Player player) {
	}
	
}

File: src/main/java/io/github/paul1365972/metay/Metay.kt
package io.github.paul1365972.metay

import io.github.paul1365972.metay.datastore.MetayDataStore
import io.github.paul1365972.metay.datastore.caches.CacheDataStore
import io.github.paul1365972.metay.datastore.endpoints.file.FileChunkedDataStore
import io.github.paul1365972.metay.datastore.endpoints.file.FolderDataStore
import io.github.paul1365972.metay.datastore.endpoints.mc.PDCDataStore
import io.github.paul1365972.metay.datastore.filters.NullableDataStore
import org.bukkit.Chunk
import org.bukkit.Location
import org.bukkit.World
import org.bukkit.block.Block
import org.bukkit.entity.Entity
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataHolder
import org.bukkit.plugin.ServicePriority
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import io.github.paul1365972.metay.util.MCDataStoreUtil as MC

class Metay : JavaPlugin(), MetayService {
    // TODO Estimate good cache sizes

    private val activeDataStores = mutableListOf<Lazy<MetayDataStore<*>>>()

    // Chunk cache size of 1024 seems reasonable if we expect each player to interact with 5 chunks at a time
    // Cache size just guesstimated, as i am still not sure how this will be used
    override val blockStore: MetayDataStore<Location> by registerLazy {
        CacheDataStore<Location>(
                FileChunkedDataStore(File(dataFolder, "block"), 256, MC::toChunkKey, MC::toBlockKey),
                1024, MC::toBlockKey)
    }

    // We load max 512 super chunks at a time, just guessing
    // 10x10 view distance at 50 plays leads to 5000 chunks, so we cache about double that amount
    override val chunkStore: MetayDataStore<Chunk> by registerLazy {
        CacheDataStore<Chunk>(
                FileChunkedDataStore(File(dataFolder, "chunk"), 512, MC::toSuperChunkKey, MC::toChunkKey),
                8192, MC::toChunkKey)
    }

    // We wont have more than 128 worlds
    override val worldStore: MetayDataStore<World> by registerLazy {
        CacheDataStore<World>(
                FolderDataStore(File(dataFolder, "world"), MC::toWorldKey),
                128, World::getUID)
    }

    // Not sure how this will be used either, so just going with 50 player with 20 blocks each seems reasonable
    override val tileEntityStore: MetayDataStore<Block> by registerLazy {
        CacheDataStore<Block>(
                NullableDataStore(PDCDataStore(), { it.state as? PersistentDataHolder }),
                1024, { it })
    }

    // Do we need to make the entity object the key ("identity" keys) or the uuid?
    // Cache size just guesstimated, 50 players with ~100 mobs each expected
    override val entityStore: MetayDataStore<Entity> by registerLazy {
        CacheDataStore<Entity>(
                PDCDataStore(),
                8124, { it.uniqueId })
    }

    // 4096 should be enough when we expect ~50 players with their inventories
    override val itemStore: MetayDataStore<ItemStack> by registerLazy {
        CacheDataStore<ItemStack>(
                NullableDataStore(PDCDataStore(), { it.itemMeta }),
                4096)
    }

    override fun onLoad() {
        server.servicesManager.register(MetayService::class.java, this, this, ServicePriority.Normal)
    }

    override fun onDisable() {
        activeDataStores.filter { it.isInitialized() }.forEach { it.value.close() }
    }

    private fun <T : MetayDataStore<*>> registerLazy(initializer: () -> T): Lazy<T> {
        return lazy { initializer() }.also { activeDataStores += it }
    }
}
